# Deliverables:

- Submit two files that have the name: <font color = 'red'>YourLastName_Exercise_3</font> that include the following files:

 1. An **HTML file** that has your Source code and output
 2. Your **ipynb script** that has your Source code and output


# Objectives:

In this exercise, you will:

 - Perform data analysis tasks on data read from a CSV file and loaded into a DataFrame object
 - Use sqlalchemy to load data stored in a DatFrame object into sqlite database engine
 - Use sqlalchemy to connect to sqlite database engine to execute  SQL  queries


Formatting Python Code
When programming in Python, refer to Kenneth Reitz’ PEP 8: The Style Guide for Python Code:
http://pep8.org/ (Links to an external site.)Links to an external site.
There is the Google style guide for Python at
https://google.github.io/styleguide/pyguide.html (Links to an external site.)Links to an external site.
Comment often and in detail.


In [1]:
# import libraries
import os
import pandas as pd 
import numpy as np  


In [2]:
# allow muliple output in one cell window
from IPython.core.interactiveshell import InteractiveShell
InteractiveShell.ast_node_interactivity = "all"

In [3]:
# read in the file
xyzcust10 = pd.read_csv('xyzcust10.csv')

# what are the data types
(xyzcust10).dtypes

ACCTNO                    object
ZIP                        int64
ZIP4                       int64
LTD_SALES                float64
LTD_TRANSACTIONS           int64
YTD_SALES_2009           float64
YTD_TRANSACTIONS_2009      int64
CHANNEL_ACQUISITION       object
BUYER_STATUS              object
ZIP9_Supercode             int64
ZIP9_SUPERCODE             int64
dtype: object

In [4]:
# info compared to dtypes
xyzcust10.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 30471 entries, 0 to 30470
Data columns (total 11 columns):
ACCTNO                   30471 non-null object
ZIP                      30471 non-null int64
ZIP4                     30471 non-null int64
LTD_SALES                30471 non-null float64
LTD_TRANSACTIONS         30471 non-null int64
YTD_SALES_2009           30471 non-null float64
YTD_TRANSACTIONS_2009    30471 non-null int64
CHANNEL_ACQUISITION      30471 non-null object
BUYER_STATUS             30471 non-null object
ZIP9_Supercode           30471 non-null int64
ZIP9_SUPERCODE           30471 non-null int64
dtypes: float64(2), int64(6), object(3)
memory usage: 2.6+ MB


In [5]:
# all objects in python are given a unique id
id(xyzcust10)

# in this case the object 'test' is the same as object 'xyzcust10'
# a change to either one changes both
test = xyzcust10
id(test)

4356156240

4356156240

In [6]:
# want to see if these two objects are the same?
# we added a column to test dataframe and it also added to xyzcust10!!!!
# that is becuase they are the same object
test['test'] = test['LTD_SALES']/test['LTD_TRANSACTIONS']
test.info()
xyzcust10.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 30471 entries, 0 to 30470
Data columns (total 12 columns):
ACCTNO                   30471 non-null object
ZIP                      30471 non-null int64
ZIP4                     30471 non-null int64
LTD_SALES                30471 non-null float64
LTD_TRANSACTIONS         30471 non-null int64
YTD_SALES_2009           30471 non-null float64
YTD_TRANSACTIONS_2009    30471 non-null int64
CHANNEL_ACQUISITION      30471 non-null object
BUYER_STATUS             30471 non-null object
ZIP9_Supercode           30471 non-null int64
ZIP9_SUPERCODE           30471 non-null int64
test                     30471 non-null float64
dtypes: float64(3), int64(6), object(3)
memory usage: 2.8+ MB
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 30471 entries, 0 to 30470
Data columns (total 12 columns):
ACCTNO                   30471 non-null object
ZIP                      30471 non-null int64
ZIP4                     30471 non-null int64
LTD_SALES           

In [7]:
# make a copy and get differnt object ids
# now you can independently manipulate each copy and leave the original intact
xyzcust10red = xyzcust10.copy() 

xyzcust10rev1 = xyzcust10.copy() 

id(xyzcust10)
id(xyzcust10red)
id(xyzcust10rev1)

4356156240

4380787856

4370329936

In [8]:
# working with a Pandas DataFrame
type(xyzcust10)

pandas.core.frame.DataFrame

In [9]:
# check out the first five records
xyzcust10.head()

# check out the last five records
xyzcust10.tail()

ACCTNO    ZIP  ZIP4  LTD_SALES  LTD_TRANSACTIONS  YTD_SALES_2009  \
0  WDQQLLDQL  60084  5016       90.0                 1             0.0   
1  WQWAYHYLA  60091  1750     4227.0                 9          1263.0   
2  GSHAPLHAW  60067   900      420.0                 3           129.0   
3  PGGYDYWAD  60068  3838     6552.0                 6             0.0   
4  LWPSGPLLS  60090  3932      189.0                 3            72.0   

   YTD_TRANSACTIONS_2009 CHANNEL_ACQUISITION BUYER_STATUS  ZIP9_Supercode  \
0                      0                  IB     INACTIVE       600845016   
1                      3                  RT       ACTIVE       600911750   
2                      1                  RT       ACTIVE       600670900   
3                      0                  RT     INACTIVE       600683838   
4                      1                  RT       ACTIVE       600903932   

   ZIP9_SUPERCODE         test  
0       600845016    90.000000  
1       600911750   469.666667  
2       600670900   140.000000  
3       600683838  1092.000000  
4       600903932    63.000000

ACCTNO    ZIP  ZIP4  LTD_SALES  LTD_TRANSACTIONS  YTD_SALES_2009  \
30466  SYDQYLSWH  60098  3951     2736.0                 9            96.0   
30467  SAPDQHQLP  60098  9681     2412.0                 8           108.0   
30468  SASYAPDSY  60098     0      429.0                 1             0.0   
30469   PWQPDWHA  60098  7927      651.0                 1             0.0   
30470  SQQHDYHWH  60098  4160     4527.0                16           672.0   

       YTD_TRANSACTIONS_2009 CHANNEL_ACQUISITION BUYER_STATUS  ZIP9_Supercode  \
30466                      1                  RT       ACTIVE       600983951   
30467                      1                  RT       ACTIVE       600989681   
30468                      0                  RT     INACTIVE       600983858   
30469                      0                  RT     INACTIVE       600987927   
30470                      2                  RT       ACTIVE       600984160   

       ZIP9_SUPERCODE      test  
30466       600983951  304.0000  
30467       600989681  301.5000  
30468       600983858  429.0000  
30469       600987927  651.0000  
30470       600984160  282.9375

xyzcust10 appears to have two nine-digit ZIP “supercode” columns with slightly different column labels or names.  To see them, try entering xyzcust10.columns or xyzcust10.dtypes at the command prompt.  Are the values in these two columns the same?  If so, we can get rid of one of them.  There are different ways we can figure out whether they are the same, but a simple way is to test each pair of values to see if they are equal or not.


In [11]:
# check out the columns
xyzcust10.columns

# can put columns in a list - may come in handy sometime
col_list = list(xyzcust10.columns)
type(col_list)
col_list

Index(['ACCTNO', 'ZIP', 'ZIP4', 'LTD_SALES', 'LTD_TRANSACTIONS',
       'YTD_SALES_2009', 'YTD_TRANSACTIONS_2009', 'CHANNEL_ACQUISITION',
       'BUYER_STATUS', 'ZIP9_Supercode', 'ZIP9_SUPERCODE', 'test'],
      dtype='object')

list

['ACCTNO',
 'ZIP',
 'ZIP4',
 'LTD_SALES',
 'LTD_TRANSACTIONS',
 'YTD_SALES_2009',
 'YTD_TRANSACTIONS_2009',
 'CHANNEL_ACQUISITION',
 'BUYER_STATUS',
 'ZIP9_Supercode',
 'ZIP9_SUPERCODE',
 'test']

In [12]:
# are the two zip code columns exactly the same
# logical comparison
(xyzcust10.ZIP9_Supercode != xyzcust10.ZIP9_SUPERCODE).sum()

0

So, Oops!  Someone included the same column in the data twice, but with slightly different names. Why waste the space?  Why risk confusion?  Let's get rid of one of them:

We could do:


In [13]:
# deletes the series in-place. Can only operate on one item at a time
del xyzcust10['ZIP9_Supercode']
del xyzcust10['test']

del xyzcust10rev1['ZIP9_Supercode']
del xyzcust10rev1['test']

or

In [16]:
# Pandas drop command
xyzcust10red.drop(['ZIP9_Supercode','test'], axis=1, inplace=True)
xyzcust10red.info()

KeyError: "['ZIP9_Supercode' 'test'] not found in axis"

Let's next look at xyz transactional data that is in a table in a SQLite3 relational database (“RDB”) file that's called xyz.db.  You should have put the xyz.db file you downloaded from Canvas in the same directory as your python code.

You can use os.getcwd() to see your current directory.

In [17]:
os.getcwd()

'/Users/gurjy/Downloads/Exercise 3 2020 version 1'

If you installed the sqlite3 client, you can take a look at this database (“DB”) without using Python. sqlSQLite3 is a very simple and easy to use RDB, and it doesn't require a server.  Assuming that you've installed it and that you're in the directory where you put xyz.db, using the command from your OS command prompt:

c:\Users\Sharon\Desktop\MSDS 420\Mod 5 Exercise 3>sqlite3 xyz.db
SQLite version 3.21.0 2017-10-24 18:55:49
Enter ".help" for usage hints.
sqlite>

will start sqlite3 and open the db file.  You can see the tables in this db with the sqlite3 command .tables .  (That's a period, “.” before tables. “Help” in sqlSQLite3 is .help  .)

sqlite> .tables
xyztrans
sqlite> 


There are a couple of different ways to read and write data to RDBs using Python, but the most flexible and easiest may be by using what's in pandas.  Pandas will make use of the SQLAlchemy package. SQLAlchemy provides a consistent interface with different RDBs, SQLite being one of them.

Let's get SQLAlchemy into our IPython session:


In [18]:
import sqlalchemy
from sqlalchemy import create_engine

Now let's specify the xyz db as the SQLite3 RDB we want to work with:

In [19]:
engine = create_engine('sqlite:///xyz.db')

This assumes that you have xyz.db in your current working directory.  There are different valid syntaxes, e.g.

sqlite:///:memory: (or, sqlite://) 
sqlite:///relative/path/to/file.db 
sqlite:////absolute/path/to/file.db 

We used the second syntax, above.  Be sure to use the correct number of slashes for the version you want to use.  You need the enclosing single quotes, too.


There's only one table in this RDB.  It's called “xyztrans.” Let's read it into a DataFrame:

In [20]:
# read in one table
xyztrans = pd.read_sql('xyztrans', engine)
xyztrans.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 62395 entries, 0 to 62394
Data columns (total 9 columns):
index           62395 non-null int64
ACCTNO          62395 non-null object
QTY             62395 non-null int64
TRANDATE        62395 non-null object
TRAN_CHANNEL    62395 non-null object
PRICE           62395 non-null float64
TOTAMT          62395 non-null float64
ORDERNO         62395 non-null object
DEPTDESCR       62395 non-null object
dtypes: float64(2), int64(2), object(5)
memory usage: 4.3+ MB


This db has only one table in it.  What if it had more than one, and you didn't know their names?  How would you know? Well, one way is to read some “metadata” from it:

In [21]:
from sqlalchemy import schema

In [22]:
xyzMetaData=schema.MetaData(bind=engine)
xyzMetaData.reflect()

In [23]:
xyzMetaData.tables

immutabledict({'xyztrans': Table('xyztrans', MetaData(bind=Engine(sqlite:///xyz.db)), Column('index', BIGINT(), table=<xyztrans>), Column('ACCTNO', TEXT(), table=<xyztrans>), Column('QTY', BIGINT(), table=<xyztrans>), Column('TRANDATE', TEXT(), table=<xyztrans>), Column('TRAN_CHANNEL', TEXT(), table=<xyztrans>), Column('PRICE', FLOAT(), table=<xyztrans>), Column('TOTAMT', FLOAT(), table=<xyztrans>), Column('ORDERNO', TEXT(), table=<xyztrans>), Column('DEPTDESCR', TEXT(), table=<xyztrans>), schema=None)})


xyzMetaData.tables will be a dict that contains information about the db.  Tables will be keys in this dict:


In [ ]:
# ----------------------------------------------------------------------------------
# if you have run this code before, you will have xyzcust table in your database
# to remove the table, you can run the code below
# ----------------------------------------------------------------------------------
sql = ('DROP TABLE IF EXISTS xyzcust;')
result = engine.execute(sql)

# you should only be able to see the yxztrans table
xyzMetaData=schema.MetaData(bind=engine)
xyzMetaData.reflect()
xyzMetaData.tables

In [ ]:
xyzMetaData.tables.keys()


At this point there's only one table name, 'xyztrans,” in xyz.db.  You'll see another method for inspecting DB's below.

We're going to write the xyz customer records into a new table in the sqlite3 RDB, but before we do that let's make sure that the records are unique, that is, that no customer has more than one record.   We can do this with some pandas DataFrame methods.  Using the customer DataFrame xyzcust10rev1


In [26]:
# lets check for row duplicates
xyzcust10rev1.duplicated().sum()

292

will return a zero if all records are unique, or the number of rows in xyzcust10rev1 that are duplicates.  The reason is that the duplicated() method for the DataFrame returns a Series of Trues and Falses, a Boolean Series.  Summing over the Series forces the values to be cast as numeric.

Oops.  There are some duplicates.  How many duplicates do you find in xyzcust10rev1? 


To rid a DataFrame of unduplicated rows, 

In [27]:
# drop duplicates and then make sure they are gone
xyzcustUnDup = xyzcust10rev1.drop_duplicates()

xyzcustUnDup.duplicated().sum()

id(xyzcust10rev1)
# a drop_duplicates creates a new object
id(xyzcustUnDup)

0

4370329936

4644505808


How many unique customer records do you now have?  By the way, note that you could have limited your examination to just one or more columns, for example just ACCTNO, customer account number, by providing ACCTNO as an argument or by using it to define a Series:


In [28]:
xyzcust10rev1.duplicated('ACCTNO').sum()

292

In [29]:
xyzcust10rev1.ACCTNO.duplicated().sum()

292


When there are duplicates of a record, which of them do you think .drop_duplicates() retains?

Now that we've checked for, and have removed, duplicate customer records, from the customer records, let's write them into a new table in xyztrans.db. 


In [30]:
# add the data from the csv file into a new table in xyz database
xyzcustUnDup.to_sql('xyzcust', engine)

Did it create the table in xyz.db?  Check:

In [31]:
pd.read_sql_table('xyzcust', engine).columns

Index(['index', 'ACCTNO', 'ZIP', 'ZIP4', 'LTD_SALES', 'LTD_TRANSACTIONS',
       'YTD_SALES_2009', 'YTD_TRANSACTIONS_2009', 'CHANNEL_ACQUISITION',
       'BUYER_STATUS', 'ZIP9_SUPERCODE'],
      dtype='object')

should produce the columns of the DataFrame you wrote to the db.  Remember that “engine” refers to the SQLite3 DB by way of defining the connection using SQLAlchemy's create_engine method.

How many tables are there now in xyz.db?  And, what are their names? 


In [32]:
xyzMetaData=schema.MetaData(bind=engine)
xyzMetaData.reflect()
xyzMetaData.tables

immutabledict({'xyzcust': Table('xyzcust', MetaData(bind=Engine(sqlite:///xyz.db)), Column('index', BIGINT(), table=<xyzcust>), Column('ACCTNO', TEXT(), table=<xyzcust>), Column('ZIP', BIGINT(), table=<xyzcust>), Column('ZIP4', BIGINT(), table=<xyzcust>), Column('LTD_SALES', FLOAT(), table=<xyzcust>), Column('LTD_TRANSACTIONS', BIGINT(), table=<xyzcust>), Column('YTD_SALES_2009', FLOAT(), table=<xyzcust>), Column('YTD_TRANSACTIONS_2009', BIGINT(), table=<xyzcust>), Column('CHANNEL_ACQUISITION', TEXT(), table=<xyzcust>), Column('BUYER_STATUS', TEXT(), table=<xyzcust>), Column('ZIP9_SUPERCODE', BIGINT(), table=<xyzcust>), schema=None), 'xyztrans': Table('xyztrans', MetaData(bind=Engine(sqlite:///xyz.db)), Column('index', BIGINT(), table=<xyztrans>), Column('ACCTNO', TEXT(), table=<xyztrans>), Column('QTY', BIGINT(), table=<xyztrans>), Column('TRANDATE', TEXT(), table=<xyztrans>), Column('TRAN_CHANNEL', TEXT(), table=<xyztrans>), Column('PRICE', FLOAT(), table=<xyztrans>), Column('TOTAMT'

In [33]:
xyzMetaData.tables.keys()

dict_keys(['xyzcust', 'xyztrans'])

Another way to look at the metadata of an RDB using SQLAlchemy is by using the “inspect” method:

In [34]:
xyzMetaData

MetaData(bind=Engine(sqlite:///xyz.db))

In [35]:
from sqlalchemy import inspect
insp = inspect(engine)
insp.get_table_names()

['xyzcust', 'xyztrans']

You can use SQLAlchemy to query a DB so as to import selected records from an RDB.  You can also append records to existing tables in an RDB, create various kinds of DB indexes, and pretty much do everything you would do using standard SQL while interacting with an RDB using a client for it.   As a query example, suppose we wanted to select from the xyz tranaction data in the xyztrans.db all transactions made in XYZ's retail stores.  These are coded as RT in the table's TRAN_CHANNEL.  We could do:


In [36]:
rttrans = pd.read_sql_query("SELECT * FROM xyztrans WHERE TRAN_CHANNEL='RT'", engine)
rttrans.sample(10)

index     ACCTNO  QTY   TRANDATE TRAN_CHANNEL   PRICE  TOTAMT  \
47554  54962  GQDQSYQYY    1  23MAR2009           RT   28.50   28.50   
17099  20241  WHPSWLWSL    1  24MAR2009           RT   59.85   59.85   
34606  40084  SYGSHGLGQ    1  05SEP2009           RT   60.00   60.00   
47190  54574  GQWHSWDHD    1  11DEC2009           RT   89.85   89.85   
42269  48906  LWHAHLQGW    1  12JUN2009           RT  149.85  149.85   
8132    9771   DHHLDHGL    1  11NOV2009           RT   45.00   45.00   
23337  27334  APLLSLSHA    1  19JUL2009           RT   20.97   20.97   
27967  32615  SQPHHHQLL    1  15OCT2009           RT   29.85   29.85   
18178  21495  WYGADAHLW    1  11DEC2009           RT   48.00   48.00   
17534  20748  WHYQYALHP    1  25AUG2009           RT  135.00  135.00   

            ORDERNO                      DEPTDESCR  
47554  CCXUKRVXXXVI  Mobile Electronic Accessories  
17099  CCXVVINXXXVI  Mobile Electronic Accessories  
34606  CCXVIIZXXXVI           Portable Electronics  
47190  CCXUZXVXXXNI           Portable Electronics  
42269  CCXIEKXXXXRC               Small Appliances  
8132   CCXNUNVXXXNX             Mobile Electronics  
23337  CCXIKXEXXXKC               Small Appliances  
27967  CCXKIXNXXXKI               Small Appliances  
18178  CCXXKCVXXXNI               Small Appliances  
17534  CCXKKKKXXXKI                     Home Audio


A last point about SQLAlchemy:  it has its own declarative language that provides means of interacting with DB's that is more “object oriented” than traditional SQL is.  You can find lots of documentation about SQLAlchemy at http://www.sqlalchemy.org.


In [37]:
# read in all customer data
customers = pd.read_sql_query("SELECT * FROM xyzcust", engine)
# look at first 5 rows
customers.head()

index     ACCTNO    ZIP  ZIP4  LTD_SALES  LTD_TRANSACTIONS  YTD_SALES_2009  \
0      0  WDQQLLDQL  60084  5016       90.0                 1             0.0   
1      1  WQWAYHYLA  60091  1750     4227.0                 9          1263.0   
2      2  GSHAPLHAW  60067   900      420.0                 3           129.0   
3      3  PGGYDYWAD  60068  3838     6552.0                 6             0.0   
4      4  LWPSGPLLS  60090  3932      189.0                 3            72.0   

   YTD_TRANSACTIONS_2009 CHANNEL_ACQUISITION BUYER_STATUS  ZIP9_SUPERCODE  
0                      0                  IB     INACTIVE       600845016  
1                      3                  RT       ACTIVE       600911750  
2                      1                  RT       ACTIVE       600670900  
3                      0                  RT     INACTIVE       600683838  
4                      1                  RT       ACTIVE       600903932

In [38]:
# read in all transactional data
transactions = pd.read_sql_query("SELECT * FROM xyztrans", engine)
# look at first five rows
transactions.head()

index  ACCTNO  QTY   TRANDATE TRAN_CHANNEL   PRICE  TOTAMT       ORDERNO  \
0      0  WGDQLA    1  09JUN2009           RT  599.85  599.85  CCXXNNXXXXUX   
1      1  WGDQLA    1  09JUN2009           RT   39.00   39.00  CCXXNNXXXXUX   
2      2  WGDQLA    1  28NOV2009           RT   15.00   15.00  CCXNXXKXXXRI   
3      3  WGDQLA    1  28NOV2009           RT   69.00   69.00  CCXNXXKXXXRI   
4      4  WGDQLA    1  28NOV2009           RT   84.00   84.00  CCXNXXKXXXRI   

          DEPTDESCR  
0        Home Audio  
1  Small Appliances  
2  Small Appliances  
3  Small Appliances  
4  Small Appliances

In [39]:
# now that you have read in your database into dataframes, you can use 
# the dataframes to analyze your data.

# BUT FOR YOUR HOMEWORK, YOU ARE TO USE SQL 

transactions['TRAN_CHANNEL'].value_counts()

transactions['DEPTDESCR'].value_counts()

customers['ZIP'].value_counts()

RT    53811
IB     6373
CB     2211
Name: TRAN_CHANNEL, dtype: int64

Mobile Electronic Accessories    18007
Small Appliances                 15641
Mobile Electronics               13063
Home Audio                        7257
Portable Electronics              5691
Cameras & Camcorder Accessori     2732
Appliances                           4
Name: DEPTDESCR, dtype: int64

60091    3422
60093    3152
60062    3059
60067    3022
60068    2754
60089    1990
60056    1515
60074    1297
60060    1283
60061    1197
60076    1081
60069     776
60077     735
60084     717
60073     678
60090     644
60098     562
60070     458
60085     377
60083     342
60081     318
60087     264
60097     150
60096     122
60071      97
60064      41
60072      34
60088      28
60078      25
60065      21
60075       5
60094       4
60082       3
60079       2
60192       2
60095       1
0           1
Name: ZIP, dtype: int64

In [40]:
# SQL groupby lets you summarize 
# https://www.sqlitetutorial.net/sqlite-group-by/

channel = pd.read_sql_query("SELECT TRAN_CHANNEL, count(ACCTNO) as Customers  \
                             FROM xyztrans \
                             GROUP BY TRAN_CHANNEL", engine)
channel

TRAN_CHANNEL  Customers
0           CB       2211
1           IB       6373
2           RT      53811

<div class="alert alert-block alert-warning">

# Requirements :

### Note that all of your answers should use SQL and SQLAlchemy to answer the homework requirements.

</div>

### Requirement 1 (6 points)

Show all customers that have over 10,000 sales in 2009 and over 25,000 sales in LTD_Sales.

In [44]:
topcust = pd.read_sql_query("SELECT * \
                            FROM xyzcust\ 
                            WHERE LTD_SALES > 25000 AND YTD_SALES_2009 > 10000", engine)
topcust.head()

index     ACCTNO    ZIP  ZIP4  LTD_SALES  LTD_TRANSACTIONS  YTD_SALES_2009  \
0   1510  LGYYAGQQA  60067  6613    47307.0                48         12684.0   
1   1872  LLSYGYDAY  60074     0  2247750.0                84        351000.0   
2  17695    SAPYPAP  60093  1320   122352.0                34         28002.0   
3  25095  SDAPWQHLD  60093  3939    40329.0                25         11616.0   

   YTD_TRANSACTIONS_2009 CHANNEL_ACQUISITION BUYER_STATUS  ZIP9_SUPERCODE  
0                      9                  RT       ACTIVE       600676613  
1                     19                  CB       ACTIVE       600746495  
2                      5                  CB       ACTIVE       600931320  
3                      7                  CB       ACTIVE       600933939

### Requirement 2 (6 points)

Show a list of customers that live in zip 60076 and have over 30 LTD Transactions

In [49]:
zipcust = pd.read_sql_query("SELECT * FROM xyzcust\
                            WHERE LTD_TRANSACTIONS > 30 AND ZIP = 60076", engine)
zipcust.head()

index     ACCTNO    ZIP  ZIP4  LTD_SALES  LTD_TRANSACTIONS  YTD_SALES_2009  \
0   3118  PHQQYAYQW  60076  3527     7407.0                31          4953.0   
1  16386  GGGSDSLQD  60076  1741    18726.0                32          2385.0   
2  22060  SPSAQHAGA  60076  1464     6534.0                35          1032.0   
3  29323  WWPYQDQDY  60076  1946     2076.0                33             0.0   

   YTD_TRANSACTIONS_2009 CHANNEL_ACQUISITION BUYER_STATUS  ZIP9_SUPERCODE  
0                     18                  RT       ACTIVE       600763527  
1                      5                  IB       ACTIVE       600761741  
2                      7                  IB       ACTIVE       600761464  
3                      0                  RT       LAPSED       600761946

### Requirement 3 (6 points)

List the inactive customers that had over 5,000 LTD sales. Sort by LTD Sales with highest amount on top.

In [59]:
inactcust = pd.read_sql_query("SELECT * \
                              FROM xyzcust \
                              WHERE LTD_SALES > 6000 AND BUYER_STATUS = 'INACTIVE' \
                              ORDER BY LTD_SALES DESC", engine)
inactcust.head(2000)

index     ACCTNO    ZIP  ZIP4  LTD_SALES  LTD_TRANSACTIONS  \
0   11714  ALGHWHDDA  60093  2322    30696.0                 3   
1    5761  WDQPGQAPG  60077     0    14946.0                 5   
2   15700  PLSPDWDSY  60093  1820    12078.0                 8   
3   28839  LQLAHQASW  60077  1081    11529.0                 5   
4   13790  GQDQPHGYS  60097  8451    10455.0                 1   
5   11095  ALDLSSASQ  60062  5500    10095.0                 4   
6    9375  GQQPYGPQW  60062     0     8889.0                 1   
7    5692  LAQSDQYYW  60061     0     8745.0                 5   
8   12919  LPDWWHDDA  60093     0     8661.0                12   
9    6535  AAYLYDYYS  60083  9607     8598.0                16   
10  28039  PWLYYSADH  60084  1961     8178.0                 7   
11   7332  LSHHGGYGH  60067  7801     8022.0                39   
12   6866  PWAYSDSSP  60091  2925     8007.0                 2   
13   7209  AGGAGDDGY  60068  1102     7980.0                 8   
14   9867  LAWPAAQLL  60061  3817     7806.0                 2   
15  20327   HGDPQQAA  60056     0     7569.0                15   
16  29121  PWLYGLHPQ  60062  5502     7308.0                10   
17   3658  LQDSGYWAA  60067  5930     7263.0                 6   
18  27009  WHALPDQDD  60062  5218     7251.0                 8   
19  27306  WYWGQSWSL  60067  2088     7176.0                 6   
20  17198  LHAPWQDHL  60068  3036     7164.0                 8   
21  17802  LQYLAPHGG  60093     0     7008.0                11   
22   5227  SLLWSYPAW  60077  1991     6882.0                 5   
23  14038  LGWQHPLPS  60062  1243     6633.0                 3   
24      3  PGGYDYWAD  60068  3838     6552.0                 6   
25  25015  LDLLHHDWH  60068  1452     6540.0                 1   
26   4314  AALHDLWAD  60062  8108     6516.0                 5   
27  30091  WPHYGHSGQ  60068  4344     6501.0                 8   
28   1052  LGWQSDHPP  60067  7736     6231.0                 2   
29  22011  AHLGPWASW  60093  4251     6186.0                14   
30   4381  AGDDPGHWW  60062  6348     6042.0                 2   
31  16342   WGGLDYGP  60060  3327     6018.0                 4   

    YTD_SALES_2009  YTD_TRANSACTIONS_2009 CHANNEL_ACQUISITION BUYER_STATUS  \
0              0.0                      0                  RT     INACTIVE   
1              0.0                      0                  RT     INACTIVE   
2              0.0                      0                  CB     INACTIVE   
3              0.0                      0                  IB     INACTIVE   
4              0.0                      0                  RT     INACTIVE   
5              0.0                      0                  RT     INACTIVE   
6              0.0                      0                  CB     INACTIVE   
7              0.0                      0                  RT     INACTIVE   
8              0.0                      0                  IB     INACTIVE   
9              0.0                      0                  IB     INACTIVE   
10             0.0                      0                  IB     INACTIVE   
11             0.0                      0                  IB     INACTIVE   
12             0.0                      0                  RT     INACTIVE   
13             0.0                      0                  CB     INACTIVE   
14             0.0                      0                  IB     INACTIVE   
15             0.0                      0                  RT     INACTIVE   
16             0.0                      0                  RT     INACTIVE   
17             0.0                      0                  RT     INACTIVE   
18             0.0                      0                  RT     INACTIVE   
19             0.0                      0                  RT     INACTIVE   
20             0.0                      0                  RT     INACTIVE   
21             0.0                      0                  RT     INACTIVE   
22             0.0             

### Requirement 4 (6 points)

Using SQL, get the same results as you would get from the following python command: <br>  transactions['DEPTDESCR'].value_counts()

In [66]:
transactions['DEPTDESCR'].value_counts()
inactcust = pd.read_sql_query("SELECT DEPTDESCR, COUNT(*) \
                              FROM xyztrans \
                              GROUP BY DEPTDESCR \
                              ORDER BY COUNT(*) DESC ", engine)
inactcust.head(7)

Mobile Electronic Accessories    18007
Small Appliances                 15641
Mobile Electronics               13063
Home Audio                        7257
Portable Electronics              5691
Cameras & Camcorder Accessori     2732
Appliances                           4
Name: DEPTDESCR, dtype: int64

DEPTDESCR  COUNT(*)
0  Mobile Electronic Accessories     18007
1               Small Appliances     15641
2             Mobile Electronics     13063
3                     Home Audio      7257
4           Portable Electronics      5691
5  Cameras & Camcorder Accessori      2732
6                     Appliances         4

### Requirement 5 (6 points)

Which customers purchased over 5,000 from Home Audio (DEPTDESCR)?
Display Account number, Zip code, Total Amount (from transactions) and DEPTDESCR.

In [85]:
custAUDIO = pd.read_sql_query("SELECT trans.ACCTNO, ZIP,TOTAMT ,DEPTDESCR\
  FROM xyztrans as trans, xyzcust as cust\
  WHERE trans.ACCTNO = cust.ACCTNO AND DEPTDESCR = 'Home Audio' AND TOTAMT > 6000", engine)

custAUDIO.head()

ACCTNO    ZIP   TOTAMT   DEPTDESCR
0  ALDGDAQQA  60091  7499.85  Home Audio
1  PQGDYASGS  60091  7499.85  Home Audio
2  LHSGAAQLH  60084  7499.85  Home Audio
3  GQGDSWGYW  60093  7499.85  Home Audio